pkg/controller/registry/resolver/cache_test.go

```diff
@@ -0,0 +1,289 @@
+package resolver
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/operator-framework/operator-registry/pkg/api"
+	"github.com/operator-framework/operator-registry/pkg/client"
+)
+
+type BundleStreamStub struct {
+	Bundles []*api.Bundle
+}
+
+func (s *BundleStreamStub) Recv() (*api.Bundle, error) {
+	if len(s.Bundles) == 0 {
+		return nil, io.EOF
+	}
+	b := s.Bundles[0]
+	s.Bundles = s.Bundles[1:]
+	return b, nil
+}
+
+type RegistryClientStub struct {
+	BundleIterator *client.BundleIterator
+}
+
+func (s *RegistryClientStub) Get() (client.Interface, error) {
+	return s, nil
+}
+
+func (s *RegistryClientStub) GetBundle(ctx context.Context, packageName, channelName, csvName string) (*api.Bundle, error) {
+	return nil, nil
+}
+
+func (s *RegistryClientStub) GetBundleInPackageChannel(ctx context.Context, packageName, channelName string) (*api.Bundle, error) {
+	return nil, nil
+}
+
+func (s *RegistryClientStub) GetReplacementBundleInPackageChannel(ctx context.Context, currentName, packageName, channelName string) (*api.Bundle, error) {
+	return nil, nil
+}
+
+func (s *RegistryClientStub) GetBundleThatProvides(ctx context.Context, group, version, kind string) (*api.Bundle, error) {
+	return nil, nil
+}
+
+func (s *RegistryClientStub) ListBundles(ctx context.Context) (*client.BundleIterator, error) {
+	return s.BundleIterator, nil
+}
+
+func (s *RegistryClientStub) HealthCheck(ctx context.Context, reconnectTimeout time.Duration) (bool, error) {
+	return false, nil
+}
+
+func (s *RegistryClientStub) Close() error {
+	return nil
+}
+
+type RegistryClientProviderStub map[CatalogKey]ClientProvider
+
+func (s RegistryClientProviderStub) ClientsForNamespaces(namespaces ...string) map[CatalogKey]ClientProvider {
+	return map[CatalogKey]ClientProvider(s)
+}
+
+func TestOperatorCacheConcurrency(t *testing.T) {
```

Comment:
Have you considered marking this a Benchmark test?

Reply:
No, why? I'd be worried that it would never be run (as far as I know no one regularly runs tests with `-bench` in this repo) and the actual timings aren't important. I primarily wanted to see a reasonable number of concurrent cache reads and writes working properly with the race detector enabled.